man/jServe.1
.\" Man page generated from reStructuredText.
.
.TH "JSERVE" "1" "Oct 03, 2018" "" "jServe"
.SH NAME
jServe \- jServe Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SRC
.SS jServe module
.INDENT 0.0
.TP
.B class jServe.jServe(callbacks={}, port=9699, host=\(aqRKJMBL0191\(aq, certfile=None)
Bases: \fI\%threading.Thread\fP
.sp
Process a json request from client, route to callback function, return json to client
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fBcallbacks\fP (\fI\%dict\fP) – Dictionary to use as callback registry \fIsee add_callback()\fP
.IP \(bu 2
\fBport\fP (\fI\%int\fP) – Port number to use
.IP \(bu 2
\fBhost\fP (\fI\%str\fP) – Hostname for server
.IP \(bu 2
\fBcertfile\fP (\fI\%str\fP) – File to use for SSL \fI(if * **None*\fP \fISSL will not be used)\fP
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B output = \(aqjson | xml\(aq
default is json, type of output to return
.UNINDENT
.INDENT 7.0
.TP
.B add_callback(method, path, cfunc)
Register callback method, function should accept json object as data.
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fBmethod\fP (\fI\%str\fP) – Method in URL POST, GET, PUT, etc..
.IP \(bu 2
\fBpath\fP (\fIre\fP) – Path in URL after host ‘/.*’, regular expression
.IP \(bu 2
\fBcfunc\fP (\fIfunction\fP) – call back function refernce pointer (see default_callback() for details)
.UNINDENT
.UNINDENT
.sp
Note user defined functions need to increment counter
.UNINDENT
.INDENT 7.0
.TP
.B callback(method, path, jdata)
Find the callback based on method and path, the call it with jdata
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fBmethod\fP (\fI\%str\fP) – Method from URL, used as primary key in callbacks
.IP \(bu 2
\fBpath\fP (\fI\%str\fP) – path from URL
.IP \(bu 2
\fBjdata\fP (\fIjson\fP) – 
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B default_callback(path, jdata, status=400, counter=0)
Call back to execute if one is not foundself.
.sp
\fBModel callbacks on this function, counter should be incremented\fP
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fBpath\fP (\fI\%str\fP) – path in calling URL
.IP \(bu 2
\fBjdata\fP (\fIjson\fP) – jquery data to process (required)
.IP \(bu 2
\fBstatus\fP (\fI\%int\fP) – last status
.IP \(bu 2
\fBcounter\fP (\fI\%int\fP) – How many times method path did a call back
.UNINDENT
.TP
.B Returns
Returns tupple with path, json object, status, and counter
.TP
.B Return type
path(\fI\%str\fP),json(json), status(\fI\%int\fP), counter(\fI\%int\fP)
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B http_reply(data, status=200)
Wraps json data in HTTP response
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fBdata\fP (\fI\%str\fP) – converted json string
.IP \(bu 2
\fBstatus\fP (\fI\%int\fP) – Status to return see stats table below
.UNINDENT
.TP
.B Returns
HTTPD respnose string
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B process_url(args)
Break up URL string recieved in method, path, json
.INDENT 7.0
.TP
.B Parameters
\fBargs\fP (\fI\%str\fP) – URL string
.TP
.B Returns
tupple with data to be used for callbacks
.TP
.B Return type
method(\fI\%str\fP), path(\fI\%str\fP), json(json)
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B run()
Opens connection, and creates Thread object to handle requests
.UNINDENT
.INDENT 7.0
.TP
.B run_url_thread(conn, addr)
Called from run(), process a single user request in a thread
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fBconn\fP (\fIsocket\fP) – came from accept()
.IP \(bu 2
\fBaddr\fP (\fI\%list\fP) – TCPIP address Information
.UNINDENT
.TP
.B Returns
sends response to Socket
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B sig_exit(signal, fname)
Receive signal to exit from OS
.INDENT 7.0
.TP
.B Parameters
.INDENT 7.0
.IP \(bu 2
\fBsignal\fP (\fIsignal\fP) – signal received
.IP \(bu 2
\fBfname\fP (\fIfile\fP) – File name
.UNINDENT
.TP
.B Returns
Exits program
.UNINDENT
.UNINDENT
.INDENT 7.0
.TP
.B start()
start() : will start the server, will not execute any other commands afterwards
.INDENT 7.0
.TP
.B Parameters
\fBNone\fP – 
.TP
.B Returns
None
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.IP \(bu 2
genindex
.IP \(bu 2
modindex
.IP \(bu 2
search
.UNINDENT
.SH AUTHOR
Tom Papadopoulos
.SH COPYRIGHT
2018, Tom Papadopoulos
.\" Generated by docutils manpage writer.
.


jServe.py
#!/usr/bin/python3
import socket, ssl, sys, threading, re, json
from urllib.parse import urlparse
from datetime import datetime

# Single Serve Socket Server, recieve command and return result closing client socket
# multiple connections handled via threads

PORT = 9699 # the default port number to run our server on
HOST = socket.gethostname() # the default hostname to run on

CERTFILE = '/home/tpapad01/etc/server.pem'
# CERTFILE = None
LOGFILE = '/var/log/jServe.log'

__version__ = "0.0.1"


class jServe(threading.Thread):
    '''
        Process a json request from client, route to callback function, return json to client

        Args:
            callbacks (dict): Dictionary to use as callback registry *see add_callback()*
            port(int): Port number to use
            host(str): Hostname for server
            certfile(str): File to use for SSL *(if * **None** *SSL will not be used)*

        Attributes:
            output = 'json | xml' : default is json, type of output to return
        '''


    def sig_exit(self,signal,fname):
        '''
        Receive signal to exit from OS

        Args:
            signal (signal): signal received
            fname (file): File name
        Returns:
            Exits program

        '''

        self.logger("jServ exit: received signal")

        self.server.close()
        sys.exit(signal)

    def logger (self, *msgs ):
        for msg in msgs:
            print (msg, file = self.logfile, end = '',flush = False )
            print (' ', file = self.logfile, end = '',flush = False )
        print ("", file = self.logfile, flush=True,  )


    def __init__(self, callbacks = {}, port=PORT , host=HOST, certfile = CERTFILE, logto = LOGFILE):
        import signal
        self.logto = logto
        self.logfile = open(self.logto, "a")
        signal.signal(signal.SIGINT,self.sig_exit)
        threading.Thread.__init__(self)
        self.port = port
        self.output = 'json'  # 'xml' is also supported
        self.host = host
        self.certfile = certfile
        self.callregistry = callbacks
        self.server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        if certfile is None:
            self.ssl = False
            self.logger('Not using SSL' )
        else:
            self.ssl = True
            self.logger('Certfile: ' + self.certfile )
        self.logger (datetime.now().strftime("jServe object initialized: %A %B %d %Y %I:%M %p %Z"))



    def start (self):
        '''
        start() : will start the server, will not execute any other commands afterwards

        Args:
            None
        Returns:
            None
        '''
        try:
            self.server.bind((self.host, self.port))
            self.logger (datetime.now().strftime("jServe started: %A %B %d %Y %I:%M %p %Z"))
        except socket.error:
            self.logger('Bind failed %s' % (socket.error))
            sys.exit(1)


        self.server.listen(10)
        if self.ssl:
            self.logger (self.certfile)
            self.ssl = ssl.wrap_socket(self.server, server_side = True, certfile=self.certfile, keyfile=self.certfile)
        self.run()

    def default_callback (self, path, jdata, status=400, counter=0):
        '''
        Call back to execute if one is not foundself.

        **Model callbacks on this function, counter should be incremented**

        Args:
            path (str): path in calling URL
            jdata (json): jquery data to process (required)
            status (int): last status
            counter (int): How many times method path did a call back

        Returns:
            path(str),json(json), status(int), counter(int): Returns tupple with path, json object, status, and counter
        '''
        self.logger ("Callback not registered")

        return (path, {}, 404, counter + 1)


    def add_callback(self,method,path,cfunc):
        '''
        Register callback method, function should accept json object as data.

        Args:
            method (str): Method in URL POST, GET, PUT, etc..
            path (re): Path in URL after host '/.*', regular expression
            cfunc(function):  call back function refernce pointer (see default_callback() for details)


        Note user defined functions need to increment counter

        '''

        self.logger ("Call back add: ",method,path,cfunc)

        if method not in self.callregistry:
            self.callregistry[method] = []
        self.callregistry[method].append({path : cfunc })

    def callback (self, method, path, jdata):
        '''
        Find the callback based on method and path, the call it with jdata

        Args:
            method(str): Method from URL, used as primary key in callbacks
            path(str): path from URL
            jdata(json):
        '''
        status=200
        counter=0
        if method in self.callregistry:
            for m in self.callregistry[method]:
                for rep in m.keys():
                    if re.fullmatch(rep,path,re.I):
                        self.logger ("MATCH rep = {} path = {}".format(rep, path))
                        path, jdata, status, counter = m[rep](path,jdata,status=status, counter=counter)  #return first match for now

            if counter != 0:
                return (path,jdata,status,counter)
        return self.default_callback(path,jdata,status,counter)

    def run_url_thread(self, conn, addr):
        '''
        Called from run(), process a single user request in a thread

        Args:
            conn(socket): came from accept()
            addr(list): TCPIP address Information
        Returns:
            sends response to Socket
        '''
        self.logger('Client connected with ' + addr[0] + ':' + str(addr[1]))

        try:
            data = conn.recv(4096)
            self.logger ("data =",data)
        except Exception as err:
            data=""
            self.logger ("data = NULL")
        try:
            self.logger ('Parsing URL')
            url = urlparse(data)
        except:
            self.logger ('ERROR: Could not parse URL properly\nData =')
            self.logger (data)

            conn.close
            return False
        self.logger ("URL:",url)

        method, wpath, jdata = self.process_url (url)
        m1, rjdata, rcode, m2 = self.callback (method, wpath, jdata)
        self.logger ("%s:%s => %s %s %s" % ( addr[0],str(addr[1]),method, wpath, json.dumps(jdata)))

        conn.sendall(bytes(self.http_reply(json.dumps(rjdata),rcode),'utf-8'))
        self.logger ("closing connection from: "  + addr[0] + ':' + str(addr[1] ))

        conn.close() # Close
        return True

    def run(self):
        '''
        Opens connection, and creates Thread object to handle requests

        '''
        self.logger('Waiting for connections on port %s' % (self.port))

        # We need to run a loop and create a new thread for each connection
        while True:
          try:
              if self.ssl:
                  conn, addr = self.ssl.accept()
                  self.logger ("Using SSL connection")
              else:
                  self.logger ("Using Unsecure connection")
                  conn, addr = self.server.accept()
              threading.Thread(target=self.run_url_thread, args=(conn, addr)).start()
          except Exception as err:
            self.logger (err)


    def http_reply(self,data,status=200):
        '''
        Wraps json data in HTTP response

        Args:
            data (str): converted json string
            status(int): Status to return see stats table below
        Returns:
            HTTPD respnose string
        '''
        stats={  100 : "Continue", 101 : "Switching Protocols",
                 200 : "OK", 201 : "Created", 202 : "Accepted", 203 : "Non-Authoritative Information",
                 204 : "No Content", 205 : "Reset Content", 206 : "Partial Content",
                 300 : "Multiple Choices", 301 : "Moved Permanently", 302 : "Found",
                 303 : "See Other", 304 : "Not Modified", 307 : "Temporary Redirect", 308 : "Permanent Redirect",
                 400 : "Bad Request", 401 : "Unauthorized", 403 : "Forbidden", 404 : "Not Found",
                 405 : "Method Not Allowed", 406 : "Not Acceptable", 407 : "Proxy Authentication Required",
                 408 : "Request Timeout", 409 : "Conflict", 410 : "Gone", 411 : "Length Required",
                 412 : "Precondition Failed", 413 : "Payload Too Large", 414 : "URI Too Long",
                 415 : "Unsupported Media Type", 416 : "Range Not Satisfiable", 417 : "Expectation Failed",
                 418 : "I'm a teapot", 422 : "Unprocessable Entity", 426 : "Upgrade Required",
                 428 : "Precondition Required", 429 : "Too Many Requests",
                 431 : "Request Header Fields Too Large", 451 : "Unavailable For Legal Reasons",
                 500 : "Internal Server Error", 501 : "Not Implemented", 502 : "Bad Gateway",
                 503 : "Service Unavailable", 504 : "Gateway Timeout", 505 : "HTTP Version Not Supported",
                 511 : "Network Authentication Required" }

        resp = "HTTP/1.1 {} {}\n".format(status,stats[status])
        resp += "Date: {} GMT\n".format(datetime.utcnow().strftime("%a, %d %b %Y %H:%M:%S"))
        resp += "Server: jServ.py\n"
        resp += "Cache-Control: no-cache\n"
        resp += "Content-Length: {}\n".format(len(data))
        resp += "Access-Control-Allow-Origin: *\n"  #Added to avoid CORS issue
        resp += "Content-Type: application/json; charset=utf-8\n"
        resp += "Connection: Closed\n\n"
        resp += data + "\n"
        self.logger ("Respone:")
        self.logger (resp)

        return resp


    def process_url (self,args):
        '''
        Break up URL string recieved in method, path, json

        Args:
            args(str): URL string

        Returns:
            method(str), path(str), json(json): tupple with data to be used for callbacks
        '''
        self.logger('process url')
        self.logger("ARGS =",args, args.path)

        def is_number(j):
            try:
                if float(j):
                    return True
            except:
                return False

        def join_escaped(lst,sep='"'):
            retstr = ""
            tbl = str.maketrans({sep: chr(92)+sep})
            for i in lst:
                k = i
                if (k[0] == '"' and k[-1] == '"') or (k[0] == "'" and k[-1] == "'") or (k[0] == sep and k[-1] == sep) :
                    k = k[1:-1]
                    k = k.translate(tbl)
                if i == ':' or i == ',' or i == '}' or i == '{' or i == "[" or i == "]" or is_number(i) or i == 'true':
                    j = k
                else:
                    j = sep + k + sep
                retstr =  retstr + " " + j
            return retstr

        urlst = args.path.decode('ascii').replace("["," [ ").replace("]"," ] ").replace("{"," { ").replace("}"," } ").replace(":"," : ").replace(","," , ").split()
        method = urlst[0]
        wpath = urlst[1]
        self.logger ("method = ", method, " wpath = ", wpath)

        if method == "GET":
            if args.query:
                params = args.query.decode('ascii').split()[0].replace("="," : ").replace("&"," , ")
                p = params.split(',')
                params = ''
                for i in p:
                    if ':' not in i:
                        i = i + ' : true '
                    params = params + i + ' , '
                params = params[:-3]
                params = params.split()

            else:
                params = []
        else:
            if args.query:
                ags = args.query
            else:
                ags = args.path
            try:
                urlst = ags.decode('ascii').replace("{"," { ").replace("}"," } ").replace(":"," : ").replace(","," , ").split()
                params = urlst[urlst.index("{")+1:urlst.index("}",-1)]
            except:
                params = []
        jstr = "{ " + join_escaped(params) + " }"
        self.logger (jstr)
        return (method,wpath, json.loads(jstr))

if __name__ == '__main__':
    ''' Below is an example and not called if imported as module '''
    def my_call (path, jdata, status=400, counter=0):
        jdata['more'] = 'testing'
        jdata['counter'] = counter
        return (path, jdata, status, counter + 1)

    myjs = jServe()
    myjs.add_callback("GET","/tes.*",my_call)
    myjs.add_callback("POST","/te.*",my_call)
    myjs.add_callback("POST","/t.*",my_call)
    myjs.add_callback("PUT","/t.*",my_call)
    myjs.start()
